verl/tools/mcp_base_tool.py

```diff
@@ -0,0 +1,111 @@
+
+    def get_openai_tool_schema(self) -> OpenAIFunctionToolSchema:
+        """Return the OpenAI tool schema."""
+        return self.tool_schema
+
+    async def create(self, instance_id: Optional[str] = None, **kwargs) -> str:
+        """Create a tool instance.
+
+        Args:
+            instance_id: The instance id of the tool.
+
+        Returns:
+            The instance id of the tool.
+        """
+        if instance_id is None:
+            instance_id = str(uuid4())
+        self._instance_dict[instance_id] = {
+            "response": "",
+            "reward": [],
+        }
+        return instance_id
+
+    async def _async_execute(self, instance_id, parameters):
+        err_msg = ""
+        try:
+            call_tool_result = await ClientManager.call_tool(self.name, parameters, self.timeout)
+        except ClientError as e:
+            err_msg = f"\n Tool call failed: {e}"
+        except ConnectionError as e:
+            err_msg = f"\n Connection failed: {e}"
+        except Exception as e:
+            err_msg = f"\n An unexpected error occurred: {e}"
+
+        logger.debug(f"Search result for instance {instance_id} with tool {self.name}: {call_tool_result.content}")
+        result, metadata = self.post_process(call_tool_result.content)
+        metadata["api_request_error"] += err_msg
+        return result, metadata
+
+    async def execute(self, instance_id: str, parameters: dict[str, Any], **kwargs) -> Tuple[str, float, dict]:
+        if self.name == "" or self.name is None or parameters is None:
+            error_msg = "Error: 'parameters' is missing or empty."
+            logger.error(f"[MCPTool] {error_msg} Received tool name: {self.name}, parameters: {parameters}")
+            return json.dumps({"result": error_msg}), 0.0, {}
+
+        try:
+            result_text, metadata = await self._async_execute(instance_id, parameters)
+
+            # Store results in instance dictionary
+            self._instance_dict[instance_id]["reward"].append(result_text.strip())
+
+            # Convert metadata to metrics
+            metrics = {"query_count": metadata.get("query_count", 0), "status": metadata.get("status", "unknown"), "total_results": metadata.get("total_results", 0), "api_request_error": metadata.get("api_request_error")}
+
+            return result_text, 0.0, metrics
+
+        except Exception as e:
+            error_result = json.dumps({"result": f"Search execution failed: {e}"})
+            logger.error(f"[MCPSearchTool] Execution failed: {e}")
+            return error_result, 0.0, {"error": str(e)}
+
+    async def calc_reward(self, instance_id: str, **kwargs) -> str:
+        return self._instance_dict[instance_id]["reward"]
+
+    async def release(self, instance_id: str, **kwargs) -> None:
+        if instance_id in self._instance_dict:
+            del self._instance_dict[instance_id]
+
+    def post_process(self, content: list):
```

Comment:
As part of the `execute` semantics, it integrates the tool's return list into the tool's response, which undoubtedly impacts metrics and rewards.